dieboljo/embedded-loop-station
# Feature/pan effect

## Related Issues:
* [Pan effect](https://app.asana.com/0/1203221076104575/1203349316387109)

## Main Changes:
* Implemented a pan effect for new recordings. There are several ways this could have been implemented, and several ways it could be applied. I chose to apply the pan effect to newly recorded audio, not track audio already recorded. For implementation, I used the **Constant Power Panning** "pan law", as described in [this paper](https://www.cs.cmu.edu/~music/icm-online/readings/panlaws/).
* Connected a button to toggle between _overdub_ and _replace_ modes. I used pin 41 since it was convenient, but any open digital pin can be used (the pin is specified in `config.h`).
* Added the capability to use USB input (nice for avoiding the loud record button clicks).
* Moved the USB input and output build flags to `platform.ini`. They are disabled by default.

## Steps To Test:
1. With the current configuration, the effect is controlled using knob **A2** (the left knob on the _Audio Tutorial Adapter_ board).
2. Compile and upload the program, then start the serial monitor. Currently, this is the only way to read the current pan position: as the knob is turned, the position is printed, with -0.5 denoting full pan left, 0.0 indicating center, and 0.5 indicating pan right.
3. Set the knob to the full left position, press the record button, and record a base loop.
4. Press the record button again to stop recording. The loop should immediately start playing back, panned to the left speaker.
5. Turn the knob all the way to the right, and record a new clip.
6. When finished, the clip should play back entirely through the right speaker.
7. Press record again, then, while recording, turn the knob back to the left.
8. The recorded audio should pan to the left in real time.
